### PR TITLE
providers: add Alibaba Cloud (aliyun)

### LINKED ISSUE #279: providers: add support for Alibaba Cloud
Fedora CoreOS is targeting Alibaba Cloud as a new platform: https://github.com/coreos/fedora-coreos-tracker/issues/276.

Afterburn should also support that, with the platform ID `aliyun`.

This cloud platform has a metadata endpoint, documented at https://www.alibabacloud.com/help/doc-detail/49122.htm.

### PR DESCRIPTION
**Pull request by lucab**
This adds support for Alibaba Cloud, with platform ID `aliyun`.

Closes: https://github.com/coreos/afterburn/issues/279

### CODE REVIEW

File: src/providers/aliyun/mod.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        let entries: Option<String> = self
+            .client
+            .get(retry::Raw, AliyunProvider::endpoint_for("public-keys/"))
+            .send()?;
+        let keys_list = entries.unwrap_or_default();
+
+        let mut out = BTreeSet::new();
+        for entry in keys_list.lines() {
+            let key_id = entry.trim_end_matches('/');
+            let ep = format!("public-keys/{}/openssh-key", key_id);
```
Comment (ashcrow) [inline src/providers/aliyun/mod.rs:83]: Just curious: Did you find the single key endpoint via trial and error or was it documented? A quick look at the docs and using search didn't turn it up.

**Review by ashcrow — APPROVED** [review]
Makes sense :+1:

File: src/providers/aliyun/mod.rs (hunk 1 above)
Comment (lucab) [reply src/providers/aliyun/mod.rs:83]: It is not documented so I had to do a bit of experiments.
However it is basically a 1:1 copy of the AWS logic: https://github.com/coreos/afterburn/blob/59d3c56ccefe093ec98d5aecb2579b88b7b5e9b5/src/providers/aws/mod.rs#L88
For that reason, I only documented the "duplicated entry" quirk.

**Comment by lucab** [comment]
I did some basic manual testing of this on an `aliyun` plain Debian VM, both attributes and ssh-keys logic worked there.

File: src/providers/aliyun/mod.rs (hunk 1 above)
Comment (jlebon) [reply src/providers/aliyun/mod.rs:83]: It's also "documented" in the cloud-init implementation :)
https://github.com/cloud-init/cloud-init/commit/4f8ceffb2e3a9feefcb718bda7a7f0f21ef7ab7c

**Review by jlebon — APPROVED** [review]
Looks sane to me!

**Comment by ashcrow** [comment]
@lucab is this ready for merge?

**Comment by zonggen** [comment]
LGTM! Pulled locally and tests passed (as expected)!

Another documentation on cloud-init about `AliYun` metadata: https://cloudinit.readthedocs.io/en/latest/topics/datasources/aliyun.html

**Review by zonggen — APPROVED** [review]
LGTM! Pulled locally and tests passed (as expected)!

Another documentation on cloud-init about `AliYun` metadata: https://cloudinit.readthedocs.io/en/latest/topics/datasources/aliyun.html

**Comment by zonggen** [comment]
Also I could add basic testing on attribute fetching method, much like this commit https://github.com/coreos/afterburn/pull/278/commits/103f694404290306c111d2b3d053268bb6bad367

**Review by zonggen — APPROVED** [review]

**Comment by lucab** [comment]
@zonggen good point, I added a fixup commit (and introduced `maplit`); how does it look now?

**Review by ashcrow — APPROVED** [review]
Looks even better now to me :smile:

**Comment by zonggen** [comment]
Nice! Works locally. Will update the testing to my commit too :+1:

**Pull request merged**